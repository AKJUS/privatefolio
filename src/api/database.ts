import PouchDB from "pouchdb"
import PouchDBFind from "pouchdb-find"

import { AuditLog, Balances, FileImport } from "../interfaces"

// declare module "pouchdb-core" {
//   interface Database<Content extends {} = {}> {
//     explain(query: any): Promise<any>
//   }
// }
PouchDB.plugin(PouchDBFind)

export const fileImportsDB = new PouchDB<FileImport>("file-imports")
export const auditLogsDB = new PouchDB<AuditLog>("audit-logs")
export const balancesDB = new PouchDB<Balances>("balances")

// try {
//   PouchDB.replicate("file-imports", "http://localhost:5984/file-imports", { live: true })
//   PouchDB.replicate("audit-logs", "http://localhost:5984/audit-logs", { live: true })
// } catch {
//   console.log("Error replicating database")
// }

fileImportsDB.on("indexing", function (event) {
  console.log("Indexing fileImportsDB:", event)
  // called when indexes are updated
})
auditLogsDB.on("indexing", function (event) {
  console.log("Indexing auditLogsDB:", event)
  // called when indexes are updated
})
