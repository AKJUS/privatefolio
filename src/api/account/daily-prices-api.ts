import { proxy } from "comlink"
import { AssetId, ChartData, ResolutionString, SavedPrice, Time, Timestamp } from "src/interfaces"
import { PRICE_API_PAGINATION, PRICE_APIS_META, PriceApiId } from "src/settings"
import { ProgressCallback } from "src/stores/task-store"
import { getAssetTicker } from "src/utils/assets-utils"
import { formatDate } from "src/utils/formatting-utils"
import { noop } from "src/utils/utils"

import { getAccount } from "../database"
import { validateOperation } from "../database-utils"
import { PAIR_MAPPER, PRICE_APIS, PRICE_MAPPER } from "../external/prices/providers"
import { updateAsset } from "./assets-api"

export async function indexDailyPrices(accountName: string, progress: ProgressCallback = noop) {
  const account = getAccount(accountName)

  progress([0, "Daily prices: updating index for 'timestamp'"])
  await account.dailyPricesDB.createIndex({
    index: {
      fields: ["timestamp"],
      name: "timestamp",
    },
  })
  progress([50, "Daily prices: updating index for 'assetId'"])
  await account.dailyPricesDB.createIndex({
    index: {
      fields: ["assetId", "timestamp"],
      name: "assetId",
    },
  })
}

export async function getPricesForAsset(
  accountName: string,
  assetId: string,
  timestamp?: Timestamp
) {
  if (assetId === "USDT" && !!timestamp) {
    return [{ time: timestamp / 1000, value: 1 }] as ChartData[]
  }

  const account = getAccount(accountName)
  await indexDailyPrices(accountName)

  const _req: PouchDB.Find.FindRequest<SavedPrice> = {
    selector: {
      assetId,
      timestamp: timestamp || { $exists: true },
    },
    sort: [
      {
        //
        assetId: "asc",
        // eslint-disable-next-line sort-keys-fix/sort-keys-fix
        timestamp: "asc",
      },
    ],
  }

  // console.log("ðŸ“œ LOG > findTransactions > _req:", _req)
  // const explain = await (account.dailyPricesDB as any).explain(_req)
  // console.log("ðŸ“œ LOG > findTransactions > explain:", explain.index)

  const prices = await account.dailyPricesDB.find(_req)

  return prices.docs.map((x) => x.price)
}

export async function getAssetPriceMap(
  accountName: string,
  timestamp: Timestamp = new Date().getTime()
) {
  await indexDailyPrices(accountName)
  const account = getAccount(accountName)

  const day: Timestamp = timestamp - (timestamp % 86400000)

  const _req: PouchDB.Find.FindRequest<SavedPrice> = {
    fields: ["assetId", "price", "source"],
    selector: { timestamp: day },
  }

  // console.log("ðŸ“œ LOG > findTransactions > _req:", _req)
  // const explain = await (account.dailyPricesDB as any).explain(_req)
  // console.log("ðŸ“œ LOG > findTransactions > explain:", explain.index)

  const prices = await account.dailyPricesDB.find(_req)

  return prices.docs.reduce(
    (map, x) => {
      map[x.assetId] = x.price
      return map
    },
    {
      USDT: { time: (day / 1000) as Time, value: 1 },
    } as Record<string, ChartData>
  )
}

export async function getPriceCursor(
  accountName: string,
  assetId: string
): Promise<Timestamp | undefined> {
  const account = getAccount(accountName)
  const prices = await account.dailyPricesDB.find({
    limit: 1,
    selector: {
      assetId,
      timestamp: { $exists: true },
    },
    sort: [
      {
        assetId: "desc",
        timestamp: "desc",
      },
    ],
  })

  if (prices.docs.length === 0) {
    return undefined
  }

  return prices.docs[0].timestamp
}

type FetchDailyPricesRequest = {
  assetIds?: AssetId[]
  priceApiMap?: Partial<Record<AssetId, PriceApiId>>
}

export async function fetchDailyPrices(
  accountName: string,
  request: FetchDailyPricesRequest,
  progress: ProgressCallback = noop,
  signal?: AbortSignal
) {
  const { assetIds, priceApiMap = {} } = request

  if (!assetIds) {
    throw new Error("No assetIds provided") // TODO prevent this
  }

  progress([0, `Fetching asset prices for ${assetIds.length} assets`])
  await indexDailyPrices(accountName)
  const account = getAccount(accountName)

  const now = Date.now()
  const today: Timestamp = now - (now % 86400000)

  const promises: (() => Promise<void>)[] = []

  for (let i = 1; i <= assetIds.length; i++) {
    // eslint-disable-next-line no-loop-func
    promises.push(async () => {
      const assetId = assetIds[i - 1]

      if (signal?.aborted) throw new Error(signal.reason)

      const preferredPriceApiId = priceApiMap[assetId]
      const priceApiIds = preferredPriceApiId
        ? [preferredPriceApiId]
        : (Object.keys(PRICE_APIS) as PriceApiId[])

      let since: Timestamp | undefined = await getPriceCursor(accountName, assetId)
      let until: Timestamp | undefined = today

      if (!since) since = today - 86400000 * (PRICE_API_PAGINATION - 1)

      for (const priceApiId of priceApiIds) {
        try {
          while (true) {
            const priceApi = PRICE_APIS[priceApiId]
            const priceMapper = PRICE_MAPPER[priceApiId]
            const pairMapper = PAIR_MAPPER[priceApiId]

            if (!priceApi || !priceMapper || !pairMapper) {
              throw new Error(`Price API "${priceApiId}" is not supported`)
            }

            const pair = pairMapper(assetId)

            const results = await priceApi({
              limit: PRICE_API_PAGINATION,
              pair,
              since,
              timeInterval: "1d" as ResolutionString,
              until,
            })

            if (!preferredPriceApiId && results.length > 0) {
              updateAsset(accountName, assetId, { priceApiId })
            }

            if (signal?.aborted) {
              throw new Error(signal.reason)
            }

            if (results.length === 0) {
              throw new Error(`${PRICE_APIS_META[priceApiId].name} EmptyResponse`)
            }

            const docIds: Array<{ id: string }> = []

            const documentMap = results.reduce((acc, result) => {
              const price = priceMapper(result)
              const timestamp = (price.time as number) * 1000
              const _id = `${assetId}-${timestamp}`

              acc[_id] = {
                _id,
                assetId,
                pair,
                price,
                timestamp,
              }

              docIds.push({ id: _id })
              return acc
            }, {} as Record<string, SavedPrice>)

            const { results: docsWithRevision } = await account.dailyPricesDB.bulkGet({
              docs: docIds,
            })

            const docs: SavedPrice[] = docsWithRevision.map((x) => ({
              ...documentMap[x.id],
              _rev: "ok" in x.docs[0] ? x.docs[0].ok._rev : undefined,
            }))

            const updates = await account.dailyPricesDB.bulkDocs(docs)
            validateOperation(updates)

            const start = (priceMapper(results[0]).time as number) * 1000
            const end = (priceMapper(results[results.length - 1]).time as number) * 1000

            progress([
              undefined,
              `Fetched ${getAssetTicker(assetId)} using ${
                PRICE_APIS_META[priceApiId].name
              } from ${formatDate(start)} to ${formatDate(end)}`,
            ])

            if (results.length !== PRICE_API_PAGINATION) {
              // reached listing date (genesis)
              break
            }

            until = start - 86400000
            since = start - 86400000 * PRICE_API_PAGINATION
          }
          break
        } catch (error) {
          progress([undefined, `Skipped ${getAssetTicker(assetId)}: ${error}`])
        }
      }
    })
  }

  let progressCount = 0

  await Promise.all(
    promises.map((fetchFn) =>
      fetchFn().then(() => {
        progressCount += 1
        progress([(progressCount / assetIds.length) * 100])
      })
    )
  )
}

export function subscribeToDailyPrices(callback: () => void, accountName: string) {
  const account = getAccount(accountName)

  const changesSub = account.dailyPricesDB
    .changes({
      live: true,
      since: "now",
    })
    .on("change", callback)

  return proxy(() => {
    try {
      changesSub.cancel()
    } catch {}
  })
}
