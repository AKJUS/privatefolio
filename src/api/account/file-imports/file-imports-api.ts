import { proxy } from "comlink"
import { FileImport } from "src/interfaces"
import { ProgressCallback } from "src/stores/task-store"
import { hashString, noop } from "src/utils/utils"

import { getAccount } from "../../database"
import { parseCsv } from "./csv-utils"

export async function addFileImport(
  file: File,
  progress: ProgressCallback = noop,
  accountName = "main"
) {
  const account = getAccount(accountName)
  const { name, type, lastModified, size } = file

  if (type !== "text/csv") {
    throw new Error("Error reading file: not .csv")
  }

  const timestamp = new Date().getTime()
  const _id = hashString(`${name}_${size}_${lastModified}`)

  await account.fileImportsDB.put({
    _id,
    lastModified,
    name,
    size,
    timestamp,
  } as unknown as FileImport)

  // parse file
  const text = await file.text()
  const { metadata, logs, transactions } = await parseCsv(text, _id, progress)

  // save logs
  progress([60, `Saving ${logs.length} audit logs to disk`])
  await account.auditLogsDB.bulkDocs(logs)

  // save transactions
  progress([80, `Saving ${transactions.length} transactions to disk`])
  await account.transactionsDB.bulkDocs(transactions)

  // save metadata
  const fileImport = await account.fileImportsDB.get(_id)
  fileImport.meta = metadata
  await account.fileImportsDB.put<FileImport>(fileImport)

  return { _id, metadata }
}

export async function getFileImports(accountName = "main") {
  const account = getAccount(accountName)
  const res = await account.fileImportsDB.allDocs<FileImport>({
    include_docs: true,
  })
  return res.rows.map((row) => row.doc) as FileImport[]
}

export async function removeFileImport(
  fileImport: FileImport,
  progress: ProgressCallback,
  accountName = "main"
) {
  const account = getAccount(accountName)
  // TODO consider pagination
  const logs = await account.auditLogsDB.allDocs({
    // Prefix search
    // https://pouchdb.com/api.html#batch_fetch
    endkey: `${fileImport._id}\ufff0`,
    startkey: fileImport._id,
  } as PouchDB.Core.AllDocsWithinRangeOptions)
  progress([25, `Removing ${logs.rows.length} audit logs`])

  // await Promise.all(
  //   logs.rows.map((row) =>
  //     auditLogsDB.remove({
  //       _id: row.id,
  //       _rev: row.value.rev,
  //     })
  //   )
  // )

  await account.auditLogsDB.bulkDocs(
    logs.rows.map((row) => ({ _deleted: true, _id: row.id, _rev: row.value.rev } as never))
  )
  // Transactions
  const txns = await account.transactionsDB.allDocs({
    // Prefix search
    // https://pouchdb.com/api.html#batch_fetch
    endkey: `${fileImport._id}\ufff0`,
    startkey: fileImport._id,
  } as PouchDB.Core.AllDocsWithinRangeOptions)
  progress([50, `Removing ${txns.rows.length} transactions`])

  // await Promise.all(
  //   txns.rows.map((row) =>
  //     transactionsDB.remove({
  //       _id: row.id,
  //       _rev: row.value.rev,
  //     })
  //   )
  // )

  await account.transactionsDB.bulkDocs(
    txns.rows.map((row) => ({ _deleted: true, _id: row.id, _rev: row.value.rev } as never))
  )

  const res = await account.fileImportsDB.remove(fileImport)
  return res.ok
}

export function subscribeToFileImports(callback: () => void, accountName = "main") {
  const account = getAccount(accountName)
  const changesSub = account.fileImportsDB
    .changes({
      live: true,
      since: "now",
    })
    .on("change", callback)

  return proxy(() => {
    try {
      // console.log("ðŸ“œ LOG > subscribeToFileImports > unsubbing")
      changesSub.cancel()
    } catch {}
  })
}
