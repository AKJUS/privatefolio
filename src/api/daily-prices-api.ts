import { ResolutionString, SavedPrice, Timestamp } from "../interfaces"
import { mapToChartData, queryPrices } from "./binance-price-api"
import { dailyPricesDB } from "./database"

export async function getAssetPrices(symbol: string) {
  await dailyPricesDB.createIndex({
    index: {
      fields: ["timestamp"],
      name: "timestamp",
    },
  })
  await dailyPricesDB.createIndex({
    index: {
      fields: ["symbol", "timestamp"],
      name: "symbol",
    },
  })
  const explain = await (dailyPricesDB as any).explain({
    selector: { symbol, timestamp: { $exists: true } },
    sort: [{ timestamp: "asc" }],
  })
  console.log("ðŸ“œ LOG > getAssetPrices > explain:", explain)

  const prices = await dailyPricesDB.find({
    selector: { symbol, timestamp: { $exists: true } },
    sort: [{ timestamp: "asc" }],
  })

  return prices.docs.map((x) => x.price)
}

export async function fetchAssetPrices() {
  // const symbols = $filterOptionsMap.get().symbol
  // console.log("Fetch daily asset prices:", symbols)
  const symbols = ["ETH"]
  const now = Date.now()
  const today = now - (now % 86400000)
  console.log("ðŸ“œ LOG > fetchAssetPrices > today:", today)

  for (const symbol of symbols) {
    let since: Timestamp = 0
    while (since !== today) {
      console.log("Fetching daily asset prices:", symbol, since)
      const pair = `${symbol}USDT`
      const source = "binance"

      const results = await queryPrices({
        limit: 300,
        pair,
        since,
        timeInterval: "1d" as ResolutionString,
      })
      console.log("Fetched daily asset prices:", symbol, since, results)

      const docs: SavedPrice[] = results.map((result) => ({
        _id: `${pair}-${source}-${result[0]}`,
        pair,
        price: mapToChartData(result),
        source,
        symbol,
        timestamp: result[0],
      }))

      dailyPricesDB.bulkDocs(docs)
      console.log("Saved daily asset prices:", docs)

      if (results.length > 0) {
        since = results[results.length - 1][0]
      } else {
        since = today
      }
    }
  }
}
