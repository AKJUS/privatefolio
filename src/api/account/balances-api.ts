import { noop } from "src/utils/utils"

import { Balance, BalanceMap, Timestamp } from "../../interfaces"
import { DB_OPERATION_PAGE_SIZE } from "../../settings"
import { ProgressCallback } from "../../stores/task-store"
import { getPricesForAsset } from "../core/daily-prices-api"
import { getAccount } from "../database"
import { countAuditLogs, findAuditLogs } from "./audit-logs-api"
import { getValue, setValue } from "./kv-api"

export async function getLatestBalances(accountName = "main"): Promise<Balance[]> {
  const account = getAccount(accountName)
  const balancesCursor = await getValue<Timestamp>("balancesCursor", undefined, accountName)

  try {
    const balanceMap = await account.balancesDB.get(String(balancesCursor))
    const { _id, _rev, timestamp, ...map } = balanceMap
    const balanceDocs = Object.keys(map).map((x) => ({ balance: map[x], symbol: x }))

    const balances = await Promise.all(
      balanceDocs.map(async (x) => {
        const prices = await getPricesForAsset(x.symbol, timestamp)
        const price = prices.length > 0 ? prices[0] : undefined

        return {
          _id: `${timestamp}_${x.symbol}`,
          ...x,
          price,
          value: price ? price.value * x.balance : undefined,
        }
      })
    )
    return balances
  } catch (error) {
    console.log("ðŸ“œ LOG > getLatestBalances > error:", error)
    return []
  }
}

export type GetHistoricalBalancesRequest = {
  accountName: string
  limit?: number
  symbol?: string
}

export async function getHistoricalBalances(request: GetHistoricalBalancesRequest) {
  const { symbol, limit, accountName = "main" } = request
  const account = getAccount(accountName)
  await account.balancesDB.createIndex({
    index: {
      fields: ["timestamp"],
      name: "timestamp",
    },
  })
  const balances = await account.balancesDB.find({
    fields: symbol ? [symbol, "timestamp"] : undefined,
    limit,
    selector: symbol
      ? {
          [symbol]: { $exists: true },
          timestamp: { $exists: true },
        }
      : {
          timestamp: { $exists: true },
        },
    sort: [
      symbol
        ? {
            // symbol: "desc", TODO TESTME
            timestamp: "desc",
          }
        : { timestamp: "desc" },
    ],
  })

  return balances.docs.reverse()
}

export type ComputeBalancesRequest = {
  accountName?: string
  pageSize?: number
  progress?: ProgressCallback
  signal?: AbortSignal
  until?: Timestamp
}

export async function computeBalances(request: ComputeBalancesRequest) {
  const {
    progress = noop,
    signal,
    pageSize = DB_OPERATION_PAGE_SIZE,
    until = Date.now(),
    accountName = "main",
  } = request

  const account = getAccount(accountName)
  // TODO cursor
  const count = await countAuditLogs(accountName)
  progress([0, `Computing balances from ${count} audit logs`])

  const latestBalances: BalanceMap = {
    _id: 0,
    timestamp: 0,
  }
  let historicalBalances: Record<number, BalanceMap> = {}

  let latestDay = 0

  for (let i = 0; i < count; i += pageSize) {
    if (signal?.aborted) {
      throw new Error(signal.reason)
    }

    const firstIndex = i + 1
    const lastIndex = Math.min(i + pageSize, count)

    progress([(i * 90) / count, `Processing logs ${firstIndex} to ${lastIndex}`])
    const logs = await findAuditLogs(
      {
        limit: pageSize,
        order: "asc",
        skip: i,
      },
      accountName
    )

    // progress([(i * 100) / count, `Processing logs ${firstIndex} to ${lastIndex} - fetch complete`])
    for (const log of logs) {
      const { symbol, changeN, timestamp } = log

      const nextDay = timestamp - (timestamp % 86400000)

      // fill the daily gaps
      if (latestDay !== 0) {
        const daysDiff = (nextDay - latestDay) / 86400000
        if (daysDiff > 1) {
          for (let i = 1; i < daysDiff; i++) {
            const gapDay = latestDay + i * 86400000
            historicalBalances[gapDay] = Object.assign({}, latestBalances)
          }
        }
      }

      // update balance
      if (!latestBalances[symbol]) {
        latestBalances[symbol] = 0
      }
      latestBalances[symbol] += changeN
      latestBalances.timestamp = nextDay

      // update audit log
      log.balance = latestBalances[symbol]

      // update historical balances
      if (!historicalBalances[nextDay]) {
        historicalBalances[nextDay] = Object.assign({}, latestBalances)
      } else {
        historicalBalances[nextDay][symbol] = latestBalances[symbol]
      }

      latestDay = nextDay
    }
    // progress([
    //   (i * 100) / count,
    //   `Processing logs ${firstIndex} to ${lastIndex} - compute complete`,
    // ])
    await account.auditLogsDB.bulkDocs(logs)

    //
    // progress([
    //   (i * 100) / count,
    //   `Processing logs ${firstIndex} to ${lastIndex} - audit logs updated`,
    // ])

    const balancesIds = Object.keys(historicalBalances).map((x) => ({ id: x }))
    const { results: balancesDocs } = await account.balancesDB.bulkGet({ docs: balancesIds })

    // eslint-disable-next-line no-loop-func
    const balances: BalanceMap[] = balancesDocs.map((doc) => ({
      ...historicalBalances[doc.id],
      _id: doc.id,
      _rev: "ok" in doc.docs[0] ? doc.docs[0].ok._rev : undefined,
      timestamp: Number(doc.id),
    }))
    // console.log("ComputeBalances db results", balances)
    await account.balancesDB.bulkDocs(balances)
    await setValue("balancesCursor", latestDay, accountName)
    progress([
      ((i + pageSize) * 90) / Math.max(count, pageSize),
      `Processing logs ${firstIndex} to ${lastIndex} complete`,
    ])

    // free memory, only keep last day
    historicalBalances = {
      [latestDay]: historicalBalances[latestDay],
    }
  }

  if (latestDay === 0) return

  progress([95, `Filling the balances to reach today`])
  // The balances remain the same until today
  for (let i = latestDay + 86400000; i <= until; i += 86400000) {
    historicalBalances[i] = latestBalances
    latestDay = i
  }
  const balancesIds = Object.keys(historicalBalances).map((x) => ({ id: x }))
  const { results: balancesDocs } = await account.balancesDB.bulkGet({ docs: balancesIds })
  // eslint-disable-next-line no-loop-func
  const balances: BalanceMap[] = balancesDocs.map((doc) => ({
    ...historicalBalances[doc.id],
    _id: doc.id,
    _rev: "ok" in doc.docs[0] ? doc.docs[0].ok._rev : undefined,
    timestamp: Number(doc.id),
  }))
  await account.balancesDB.bulkDocs(balances)
  await setValue("balancesCursor", latestDay, accountName)
  progress([100, `Computed all balances!`])
}
