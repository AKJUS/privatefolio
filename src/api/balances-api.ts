import { Balance, BalanceMap } from "../interfaces"
import { findAuditLogs } from "./audit-logs-api"
import { getPricesForAsset } from "./daily-prices-api"
import { auditLogsDB, balancesDB } from "./database"
import { getValue, setValue } from "./kv-api"

export async function getLatestBalances(): Promise<Balance[]> {
  const balancesCursor = await getValue("balancesCursor")

  try {
    const balanceMap = await balancesDB.get(String(balancesCursor))
    const { _id, _rev, timestamp, ...map } = balanceMap
    const balanceDocs = Object.keys(map).map((x) => ({ balance: map[x], symbol: x }))

    const balances = await Promise.all(
      balanceDocs.map(async (x) => {
        const prices = await getPricesForAsset(x.symbol, timestamp)
        const price = prices.length > 0 ? prices[0] : undefined

        return {
          ...x,
          price,
          value: price ? price.value * x.balance : undefined,
        }
      })
    )
    return balances
  } catch {
    return []
  }
}

export async function getHistoricalBalances(symbol?: string, limit = 300) {
  await balancesDB.createIndex({
    index: {
      fields: ["timestamp"],
      name: "timestamp",
    },
  })
  const balances = await balancesDB.find({
    fields: symbol ? [symbol, "timestamp"] : undefined,
    limit,
    selector: symbol
      ? {
          [symbol]: { $exists: true },
          timestamp: { $exists: true },
        }
      : {
          timestamp: { $exists: true },
        },
    sort: [
      symbol
        ? {
            symbol: "desc",
            timestamp: "desc",
          }
        : { timestamp: "desc" },
    ],
  })

  return balances.docs.reverse()
}

export async function computeBalances(symbol?: string) {
  const { length: count } = await findAuditLogs({ fields: [], filters: { symbol } })
  console.log("Recompute balances total logs:", count)

  const latestBalances: BalanceMap = {
    _id: 0,
    timestamp: 0,
  }
  let historicalBalances: Record<number, BalanceMap> = {}

  for (let i = 0; i < count; i += 500) {
    console.log("Recompute balances processing logs from", i, "to", i + 500)
    const logs = await findAuditLogs({
      filters: { symbol },
      limit: 500,
      order: "asc",
      skip: i,
    })
    console.log(
      "Recompute balances processing logs from",
      i,
      "to",
      i + 500,
      "fetch completed",
      logs
    )

    let latestDay = 0

    for (const log of logs) {
      const { symbol, changeN, timestamp } = log

      const nextDay = timestamp - (timestamp % 86400000)

      // fill the daily gaps
      if (latestDay !== 0) {
        const daysDiff = (nextDay - latestDay) / 86400000
        if (daysDiff > 1) {
          for (let i = 1; i < daysDiff; i++) {
            const gapDay = latestDay + i * 86400000
            historicalBalances[gapDay] = Object.assign({}, latestBalances)
          }
        }
      }

      // update balance
      if (!latestBalances[symbol]) {
        latestBalances[symbol] = 0
      }
      latestBalances[symbol] += changeN

      // update audit log
      log.balance = latestBalances[symbol]

      // update historical balances
      if (!historicalBalances[nextDay]) {
        historicalBalances[nextDay] = Object.assign({}, latestBalances)
      } else {
        historicalBalances[nextDay][symbol] = latestBalances[symbol]
      }

      latestDay = nextDay
    }
    console.log("Recompute balances processing logs from", i, "to", i + 500, "compute completed")
    await auditLogsDB.bulkDocs(logs)

    //
    console.log("Recompute balances processing logs from", i, "to", i + 500, "audit logs saved")
    const balancesIds = Object.keys(historicalBalances).map((x) => ({ id: x }))
    const { results: balancesDocs } = await balancesDB.bulkGet({ docs: balancesIds })
    console.log("ðŸ“œ LOG > computeBalances > balancesDocs:", balancesDocs)

    // eslint-disable-next-line no-loop-func
    const balances: BalanceMap[] = balancesDocs.map((doc) => ({
      ...historicalBalances[doc.id],
      _id: doc.id,
      _rev: "ok" in doc.docs[0] ? doc.docs[0].ok._rev : undefined,
      timestamp: Number(doc.id),
    }))
    console.log("ðŸ“œ LOG > constbalances:BalanceMap[]=Object.keys > balances:", balances)
    await balancesDB.bulkDocs(balances)
    await setValue("balancesCursor", latestDay)
    console.log("Recompute balances processing logs from", i, "to", i + 500, "balances saved")

    // free memory, only keep last day
    historicalBalances = {
      [latestDay]: historicalBalances[latestDay],
    }
  }
}
