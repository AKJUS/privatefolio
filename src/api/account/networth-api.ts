import { noop } from "src/utils/utils"

import { Networth, Time } from "../../interfaces"
import { ProgressCallback } from "../../stores/task-store"
import { getAssetPriceMap } from "../core/daily-prices-api"
import { getAccount } from "../database"
import { getHistoricalBalances } from "./balances-api"

export async function getHistoricalNetworth(accountName = "main") {
  const account = getAccount(accountName)
  const balances = await account.networthDB.allDocs({
    include_docs: true,
  })

  return balances.rows.map((x) => x.doc) as Networth[]
}

export async function computeNetworth(progress: ProgressCallback = noop, accountName = "main") {
  const account = getAccount(accountName)

  const balances = await getHistoricalBalances({
    accountName,
  })
  const count = balances.length
  progress([0, `Computing historical networth for ${count} total balances`])

  const networthArray: Networth[] = await Promise.all(
    balances.map(async ({ _id, _rev, timestamp, ...balanceMap }) => {
      const priceMap = await getAssetPriceMap(timestamp)

      const totalValue = Object.keys(priceMap).reduce((acc, symbol) => {
        const price = priceMap[symbol]
        const balance = balanceMap[symbol]

        if (!price || !balance) return acc

        return acc + Math.round(price.value * balance * 100) / 100
      }, 0)

      return {
        _id,
        change: 0,
        changePercentage: 0,
        time: (timestamp / 1000) as Time,
        value: totalValue,
      }
    })
  )

  await account.networthDB.bulkDocs(networthArray)
  progress([100, `Computed all networth records!`])
}
