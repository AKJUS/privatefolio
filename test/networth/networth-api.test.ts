import fs from "fs"
import { join } from "path"
import { computeBalances } from "src/api/account/balances-api"
import { addFileImport } from "src/api/account/file-imports/file-imports-api"
import { computeNetworth, getHistoricalNetworth } from "src/api/account/networth-api"
import { fetchDailyPrices } from "src/api/core/daily-prices-api"
import { resetAccount } from "src/api/database"
import { ProgressUpdate } from "src/stores/task-store"
import { beforeAll, expect, it } from "vitest"

const accountName = "green"

beforeAll(async () => {
  //
  await resetAccount(accountName)
  //
  const fileName = "coinmama.csv"
  const filePath = join("test/files", fileName)
  const buffer = await fs.promises.readFile(filePath, "utf8")
  const file = new File([buffer], fileName, { lastModified: 0, type: "text/csv" })
  await addFileImport(file, undefined, accountName)
  const until = Date.UTC(2017, 8, 14, 0, 0, 0, 0) // 14 Sep 2017
  await computeBalances(accountName, { until })
  await fetchDailyPrices({ priceApiId: "coinbase", symbols: ["BTC"] })
})

it.sequential("should compute historical networth", async () => {
  // arrange
  // act
  const updates: ProgressUpdate[] = []
  await computeNetworth(accountName, 0, (state) => updates.push(state))
  const networthArray = await getHistoricalNetworth(accountName)
  // assert
  expect(updates.join("\n")).toMatchInlineSnapshot(
    `
    "5,Computing networth for 14 days
    10,Computing networth starting Sep 01, 2017
    99,Setting networth cursor to Sep 14, 2017"
  `
  )
  expect(networthArray.length).toMatchInlineSnapshot(`14`)
  expect(networthArray).toMatchSnapshot()
})

it.sequential("should refresh networth", async () => {
  // arrange
  // act
  const updates: ProgressUpdate[] = []
  await computeNetworth(accountName, undefined, (state) => updates.push(state))
  const networthArray = await getHistoricalNetworth(accountName)
  // assert
  expect(updates.join("\n")).toMatchInlineSnapshot(`
    "5,Computing networth for 1 days
    10,Computing networth starting Sep 14, 2017
    99,Setting networth cursor to Sep 14, 2017"
  `)
  expect(networthArray.length).toMatchInlineSnapshot(`14`)
  expect(networthArray).toMatchSnapshot()
})
