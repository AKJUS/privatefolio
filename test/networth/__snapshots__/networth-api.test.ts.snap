// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should fetch historical networth 2`] = `
[
  {
    "_id": "1504224000000",
    "_rev": "1-1e90d2489635aa8e3a5ac4d0c0805a12",
    "change": 0,
    "changePercentage": 0,
    "time": 1504224000,
    "value": 173.96,
  },
  {
    "_id": "1504310400000",
    "_rev": "1-1701f0c7796c51088d866a515cda6981",
    "change": 0,
    "changePercentage": 0,
    "time": 1504310400,
    "value": 160.9,
  },
  {
    "_id": "1504396800000",
    "_rev": "1-2b9a4eb26388ab64ecf59183ef376311",
    "change": 0,
    "changePercentage": 0,
    "time": 1504396800,
    "value": 162.23,
  },
  {
    "_id": "1504483200000",
    "_rev": "1-d52fbf4e4a4fffc93ac9bab621088d75",
    "change": 0,
    "changePercentage": 0,
    "time": 1504483200,
    "value": 147.52,
  },
  {
    "_id": "1504569600000",
    "_rev": "1-1efa458fb8ac8361c3e9dd2516740427",
    "change": 0,
    "changePercentage": 0,
    "time": 1504569600,
    "value": 157.1,
  },
  {
    "_id": "1504656000000",
    "_rev": "1-74bfbaaa1be9fe4665654c9c59a71550",
    "change": 0,
    "changePercentage": 0,
    "time": 1504656000,
    "value": 166.22,
  },
  {
    "_id": "1504742400000",
    "_rev": "1-abd2c545dbbc6e1919295dae59e2023c",
    "change": 0,
    "changePercentage": 0,
    "time": 1504742400,
    "value": 342.29,
  },
  {
    "_id": "1504828800000",
    "_rev": "1-e07bfc059361f4a49345dad383f0ce5d",
    "change": 0,
    "changePercentage": 0,
    "time": 1504828800,
    "value": 312.47,
  },
  {
    "_id": "1504915200000",
    "_rev": "1-6988c0417ae5564d8f0a986f5a46c265",
    "change": 0,
    "changePercentage": 0,
    "time": 1504915200,
    "value": 310.72,
  },
  {
    "_id": "1505001600000",
    "_rev": "1-daf56ce189d6fdd3f3ae463bd9bf1a35",
    "change": 0,
    "changePercentage": 0,
    "time": 1505001600,
    "value": 301.35,
  },
  {
    "_id": "1505088000000",
    "_rev": "1-6b55b8611a31e92f54ab623348790b00",
    "change": 0,
    "changePercentage": 0,
    "time": 1505088000,
    "value": 307.04,
  },
  {
    "_id": "1505174400000",
    "_rev": "1-c218fdc24c06ea9de5396a5a4f0aa4bf",
    "change": 0,
    "changePercentage": 0,
    "time": 1505174400,
    "value": 303.78,
  },
  {
    "_id": "1505260800000",
    "_rev": "1-8cc3fd762e07018a5d386bcf17607397",
    "change": 0,
    "changePercentage": 0,
    "time": 1505260800,
    "value": 287.8,
  },
  {
    "_id": "1505347200000",
    "_rev": "1-29b802351f3d853f940391de1748ff82",
    "change": 0,
    "changePercentage": 0,
    "time": 1505347200,
    "value": 232.67,
  },
]
`;
